app/controllers/PaypalController.scala

```diff
@@ -248,30 +141,191 @@ class PaypalController(paymentServices: PaymentServices, checkToken: CSRFCheck,
+                              refererPageviewId: Option[String],
+                              refererUrl: Option[String],
+                              ophanPageviewId: Option[String],
+                              ophanBrowserId: Option[String],
+                              ophanVisitId: Option[String]
+                            ): AuthRequest = {
+        val safeRefererUrl = refererUrl.flatMap(url => Try(Uri.parse(url).copy(fragment = None).toString).toOption)
+        new AuthRequest(countryGroup, amount, cmp, intCmp, refererPageviewId, safeRefererUrl, ophanPageviewId, ophanBrowserId, ophanVisitId)
+      }
+
+      private implicit val countryGroupReads = new Reads[CountryGroup] {
+        override def reads(json: JsValue): JsResult[CountryGroup] = json match {
+          case JsString(id) => CountryGroup.byId(id).map(JsSuccess(_)).getOrElse(JsError("invalid CountryGroup id"))
+          case _ => JsError("invalid value for country group")
+        }
+      }
 
-    contributor.map { _ =>
-      Redirect(url, SEE_OTHER)
+      implicit val authRequestReads: Reads[AuthRequest] = (
+        (__ \ "countryGroup").read[CountryGroup] and
+          (__ \ "amount").read(min[BigDecimal](1)) and
+          (__ \ "cmp").readNullable[String] and
+          (__ \ "intCmp").readNullable[String] and
+          (__ \ "refererPageviewId").readNullable[String] and
+          (__ \ "refererUrl").readNullable[String] and
+          (__ \ "ophanPageviewId").readNullable[String] and
+          (__ \ "ophanBrowserId").readNullable[String] and
+          (__ \ "ophanVisitId").readNullable[String]
+        ) (AuthRequest.withSafeRefererUrl _)
     }
+
+    case class AuthResponse(approvalUrl: Uri, paymentId: String)
+
+    object AuthResponse {
+      import cats.syntax.either._
+      import scala.collection.JavaConverters._
+
+      def fromPayment(payment: Payment): Either[String, AuthResponse] = Either.fromOption(for {
+        links <- Option(payment.getLinks)
+        approvalLinks <- links.asScala.find(_.getRel.equalsIgnoreCase("approval_url"))
+        approvalUrl <- Option(approvalLinks.getHref)
+        paymentId <- Option(payment.getId)
+      } yield AuthResponse(Uri.parse(approvalUrl), paymentId), "Unable to parse payment")
+
+      private implicit val uriWrites = new Writes[Uri] {
+        override def writes(uri: Uri): JsValue = JsString(uri.toString)
+      }
+
+      implicit val authResponseWrites: Writes[AuthResponse] = Json.writes[AuthResponse]
+    }
+  }
+
+  // Class containing methods used to implement the execute payment endpoint.
+  // Factored out here to facilitate unit testing.
+  class ExecutePaymentUtils(paypalService: PaypalService, countryGroup: CountryGroup, cloudWatchMetrics: CloudWatchMetrics)
+    extends PaypalControllerUtils {
+
+    def executePayment(paymentId: String, payerId: String)(implicit request: ABTestRequest[AnyContent]): EitherT[Future, String, Payment] =
```

Comment:
As discussed - you could probably get rid of this one and just call `paypalService.executePayment` directly